ATSAM CAN driver (MCAN peripheral)

rleh opened the pull request:
- [x] Porting of ST's FDCAN periphal driver (based on the same 3rd party intellectual property)
- [x] Use of existing the `message_ram.hpp` implementation
- [x] Add CAN pinout to SAMV71-XPLAINED board module (there is a CAN transceiver on the dev board!)
- [x] Adapt `connect<...>()`
- [x] Update HAL matrix
- [x] Long frame and fast frame support
- [x] Remove software-buffers and make hardware-buffers directly accessible and configurable.
- [x] Test on real hardware

rleh committed:
[sam] Add support for MCAN peripheral

rleh committed:
[board] Add CAN support to SAMV71-XPLAINED board

rleh commented:
> `connect<...>()` function has some issue

The implementation:
```cpp
// main.cpp
Mcan0::connect<GpioB2::Tx, GpioB3::Rx>(InputType::PullUp);

// can_0.hpp
template<class... Pins>
static void
connect(modm::platform::InputType inputType = modm::platform::InputType::Floating)
{
	using RxPin = GetPin_t<PeripheralPin::Rx, Pins...>;
	using TxPin = GetPin_t<PeripheralPin::Tx, Pins...>;

	using Peripheral = Peripherals::Mcan<0>;

	if constexpr (!std::is_void_v<RxPin>) {
		using RxConnector = typename RxPin::template Connector<Peripheral, Peripheral::Canrx>;
		RxConnector::connect();
	}

	if constexpr (!std::is_void_v<TxPin>) {
		using TxConnector = typename TxPin::template Connector<Peripheral, Peripheral::Cantx>;
		TxConnector::connect();
	}
//...
}
```

The error:
```
modm/src/modm/platform/can/can_0.hpp: In instantiation of 'static void modm::platform::Mcan0::connect(modm::platform::InputType) [with Pins = {modm::platform::Gpio<modm::platform::gpio::B2>::As<modm::platform::PeripheralPin::Tx>, modm::platform::Gpio<modm::platform::gpio::B3>::As<modm::platform::PeripheralPin::Rx>}]':
main.cpp:32:58:   required from here
modm/src/modm/platform/can/can_0.hpp:103:10: error: type/value mismatch at argument 2 in template parameter list for 'template<class PinConfig> template<modm::platform::PeripheralPin peripheral_pin_v> template<class peripheral, class ... RequiredSignals> struct modm::platform::Gpio<_>::As<peripheral_pin_v>::Connector'
  103 |    using RxConnector = typename RxPin::template Connector<Peripheral, Peripheral::Canrx>;
      |          ^~~~~~~~~~~
modm/src/modm/platform/can/can_0.hpp:103:10: note:   expected a type, got 'Canrx'
modm/src/modm/platform/can/can_0.hpp:108:10: error: type/value mismatch at argument 2 in template parameter list for 'template<class PinConfig> template<modm::platform::PeripheralPin peripheral_pin_v> template<class peripheral, class ... RequiredSignals> struct modm::platform::Gpio<_>::As<peripheral_pin_v>::Connector'
  108 |    using TxConnector = typename TxPin::template Connector<Peripheral, Peripheral::Cantx>;
      |          ^~~~~~~~~~~
modm/src/modm/platform/can/can_0.hpp:108:10: note:   expected a type, got 'Cantx
```

@salkinium Can you tell me what I am doing wrong?

chris-durand commented:
> @salkinium Can you tell me what I am doing wrong?

I am not @salkinium, but `Canrx` and `Cantx` are templates. Try `Peripheral::Cantx<{{ id | int }}>`.

chris-durand commented on src/modm/platform/can/atsam-fdcan/can.hpp.in:
```diff
@@ -0,0 +1,387 @@
+
+#include "can_bit_timings.hpp"
+
+#define SRAMCAN_BASE (0) // TODO
```
There is an important difference regarding the message RAM compared to STM32G4. On those controllers the message RAM is a dedicated hardware SRAM exclusively reserved for the CAN peripheral.

On SAMx7x there is no such thing. The data structure resides in normal RAM and the addresses for the start of various sections are configured in the registers. The user has to allocate adequately sized 4-byte aligned blocks of memory and write their addresses to the respective registers. The size of buffers and filter counts are also fully customizable.

As a start you could just assume the fixed layout of the STM32G4 version and defined an equally sized array as the message RAM.

chris-durand replied:
That is what they are doing in the vendor HAL: https://github.com/osmocom/atmel-asf-projects/blob/3634cc8e5d3124e3cf6cf46c0f1d5385ff511426/sam/drivers/mcan/mcan.c#L65

chris-durand commented on examples/samv71_xplained_ultra/can/project.xml:
```diff
@@ -0,0 +1,10 @@
+<library>
+  <extends>modm:samv71-xplained-ultra</extends>
+  <options>
+    <option name="modm:build:build.path">../../../build/samv71_xplained_ultra/blink</option>
```
```suggestion
    <option name="modm:build:build.path">../../../build/samv71_xplained_ultra/can</option>
```

rleh commented:
> - [ ] Test on real hardware

It does not work yet; I suspect the clock configuration...

chris-durand committed:
[sam] Add PCK configuration to PMC driver and fix enableUPll()

Co-authored-by: Raphael Lehmann <raphael@rleh.de>

salkinium commented on examples/samv71_xplained_ultra/can/main.cpp:
```diff
@@ -0,0 +1,69 @@
+
+#include <modm/board.hpp>
+#include <modm/debug/logger.hpp>
+#include <modm/board.hpp>
```
```suggestion
#include <modm/board.hpp>
#include <modm/board.hpp>
#include <modm/board.hpp>
#include <modm/board.hpp>
#include <modm/board.hpp>
#include <modm/board.hpp>
```
😉

rleh replied:
Sicher ist sicher!

salkinium commented on src/modm/platform/can/atsam-fdcan/can.cpp.in:
```diff
@@ -0,0 +1,624 @@
+}
+
+void
+modm::platform::Mcan{{ id }}::initializeWithPrescaler(
```
You could use a `namespace modm::platform { }` to reduce the duplication a bit.

salkinium commented on src/modm/platform/clock/sam_pmc/clockgen.hpp.in:
```diff
@@ -13,6 +13,7 @@
 
 #pragma once
 
+#include <algorithm>
```
Move to cpp file?

salkinium reviewed (changes requested):
Very nice and clean, only some minor notes.

chris-durand committed:
[architecture] Fix compiler warnings in can message, more constexpr

Fix deprecated warning about implicitly-declared copy constructor with
custom assignment operator.

rleh commented:
The Atmel variant of this CAN IP block has configurable buffers (total size up to 4352 words == 17KB) located in the "userspace" SRAM. (STM32 has only <1KB fixed message RAM and e.g. the Rx FIFOs only fit 3 elements each.)

The buffer element sizes are also configurable to not waste memory if 64 byte CAN messages are not used (on STM32 the elements are fixed size for up to 64 bytes CAN data), we can use the [`modm:architecture:can:message.buffer` option](https://modm.io/reference/module/modm-architecture-can/#messagebuffer) to determine the necessary buffer element sizes.

Thus I suggest to adapt this driver, get rid of the software-buffers in favor of the large hardware-buffers and significantly less memory copying.

![image](https://user-images.githubusercontent.com/2820734/218153261-696a5859-31d4-4bab-864c-76a529ff2af2.png)

rleh committed:
[architecture] can: validate message buffer length and fix module dependencies

rleh commented on src/modm/platform/can/atsam-fdcan/can_interrupt.cpp.in:
```diff
@@ -0,0 +1,33 @@
+MODM_ISR(MCAN{{ id }}_INT0)
+{
+    // FIXME this does not work if the a non-default MessageRamConfig is used
+	const auto callback = Mcan{{ id }}<>::getErrorInterruptCallback();
```
Any suggestion how to access the  "static-"class with unknown template parameters from the interrupt?

chris-durand replied:
You can use a regular value type for the message RAM config and pass it to the driver as a non-type template parameter. That can be stored in a variable and accessed from the interrupt.

chris-durand commented on src/modm/platform/can/atsam-fdcan/can.hpp:
```diff
@@ -0,0 +1,433 @@
+
+#include "message_ram.hpp"
+
+namespace {
+using ErrorCallback = void (*)();
+
+extern ErrorCallback mcan0ErrorCallback;
+extern ErrorCallback mcan1ErrorCallback;
+}
```
`mcan0ErrorCallback` is defined inside namespace `modm::platform` in can_interrupt.cpp. The `extern` declarations here are in an anonymous namespaces and can't refer to the definitions in can_interrupt.cpp. If `setErrorInterruptCallback()` was actually used and instantiated the result would be a linker error. 

An anonymous namespaces gives symbols defined in it internal linkage, like `static`. Therefore, `extern` declarations in an anonymous namespace are nonsensical.

rleh committed:
[board] Fix inverted LEDs on samv71_xplained_ultra

rleh committed:
[example] Long-frame CAN example for samv71_xplained_ultra

salkinium commented on examples/samv71_xplained_ultra/can_dual/main.cpp:
```diff
@@ -32,37 +32,41 @@ class SendThread : public modm::pt::Protothread
 	size_t i;
 
-	std::chrono::microseconds delay = std::chrono::microseconds{1};
+	std::chrono::nanoseconds delay = std::chrono::nanoseconds{1};
```
That's short

salkinium commented on examples/samv71_xplained_ultra/can_dual/main.cpp:
```diff
@@ -77,6 +81,7 @@ class ReceiveThread : public modm::pt::Protothread
 		{
 			PT_WAIT_UNTIL(Can::isMessageAvailable());
 			MODM_LOG_INFO << "Mcan" << id << ": Message is available... ";
+			__DSB();
```
?

rleh replied:
Yeah I still sometimes have weird behavior with (long frame) messages reading all zero and was checking if more `__DSB();` at various places solves the problem.

salkinium commented on src/modm/platform/can/atsam-fdcan/can_impl.hpp:
```diff
@@ -0,0 +1,637 @@
+		return false;
+	}
+
+	// TODO: Is INIT mode really necessary to add/modify filters?
```
Well?

rleh replied:
I can't find any hint the in datasheet INIT mode would be necessary, and accessing the filter (located in message RAM in standard SRAM) should not cause any issues.
In the worst case some messages that should not get accepted will get accepted if the MCAN peripheral reads the filter while it is being written.

chris-durand replied:
Could you have some weird behavior when you receive a message while updating filters?

rleh replied:
The standard filter is safe because the whole config is a single 32 bit word, but the extended filter consists of two 32 bit words.
If we write the word containing the enable bit(s) to zero first, then update the other filter config word and finally write the word with the enable (==mode selection) bits, we should be safe.

rleh replied:
I fixed it in 65ec533a4.

salkinium commented on src/modm/platform/can/atsam-fdcan/message_ram.hpp:
```diff
@@ -0,0 +1,337 @@
+		// bits 28-0: can id
+	};
+	MODM_FLAGS32(CommonFifoHeader);
+	using CanId_t = Value<CommonFifoHeader_t, 29>;
```
Configuration? Value is numeric.

rleh replied:
CanId is numeric, what exactly do you mean?

![image](https://user-images.githubusercontent.com/2820734/230684680-31b807ab-af46-4b05-8e79-b73c4a937748.png)

salkinium replied:
Ah, then I got confused.

salkinium commented on src/modm/platform/can/atsam-fdcan/message_ram.hpp:
```diff
@@ -0,0 +1,337 @@
+	MODM_FLAGS32(RxFifoHeader);
+	using FilterIndex_t = Value<RxFifoHeader_t, 7, 24>;
+	using RxDlc_t = Value<RxFifoHeader_t, 4, 16>;
+	using Timestamp_t = Value<RxFifoHeader_t, 16>;
```
I think you want Configuration. And the shifts/masks look wrong to me.

salkinium replied:
This is probably right then too.

salkinium commented on src/modm/platform/can/atsam-fdcan/module.lb:
```diff
@@ -0,0 +1,75 @@
+        device = env[":target"]
+        driver = device.get_driver("mcan")
+
+        properties = device.properties
```
```suggestion
        properties = {}
```

salkinium commented on src/modm/architecture/module.lb:
```diff
@@ -121,12 +121,18 @@ class Can(Module):
+        def validate_dlc(s: int):
+            valid_dlcs = [8, 12, 16, 20, 24, 32, 48, 64]
+            if s not in valid_dlcs:
+                raise ValueError("Input must be a valid CAN data length.")
```
```suggestion
                raise ValueError(f"Input must be a valid CAN data length: {', '.join(valid_dlcs)}!")
```

rleh commented:
The pull request is now ready for reviewing and merging :)

chris-durand commented:
I'll review tomorrow.

chris-durand commented on src/modm/platform/can/sam-mcan/can.hpp:
```diff
@@ -0,0 +1,434 @@
+
+namespace {
+using ErrorCallback = void (*)();
+}
```
That should be declared somewhere inside namespace `modm`. This is effectively in global scope. Anonymous namespaces only affect linkage.

chris-durand commented on src/modm/platform/can/sam-mcan/can.hpp:
```diff
@@ -0,0 +1,434 @@
+
+	// helper for static_assert
+	template<class T = void>
+	static inline constexpr bool
```
```suggestion
	static constexpr bool
```

`constexpr` implies `inline` here.

chris-durand commented on src/modm/platform/can/sam-mcan/can.hpp:
```diff
@@ -0,0 +1,434 @@
+			using DataTimings = CanBitTiming<
+				SystemClock::Mcan,
+				fastDataBitrate,
+				5, 5, 4, 4
```
```suggestion
				5, 5, 4, 3
```
`SJW` is only 3 bits but the value is hard coded to 1 in the calculator anyway.

chris-durand commented on src/modm/platform/can/sam-mcan/can.hpp:
```diff
@@ -0,0 +1,434 @@
+	static inline uint8_t
+	getReceiveErrorCounter()
+	{
+		return ((Regs()->MCAN_ECR >> 8) & 0xFF);
```
Bit 15 is "Receive error passive". The counter is 7 bit only.

chris-durand commented on src/modm/platform/can/sam-mcan/can_impl.hpp:
```diff
@@ -0,0 +1,619 @@
+		return false;
+	}
+
+	EnterInitMode init;
```
Init mode is set here, for other filter configuration functions it isn't. Is that intended?

chris-durand commented on src/modm/platform/can/sam-mcan/can_impl.hpp:
```diff
@@ -0,0 +1,619 @@
+void
+McanDriver<id, mrc>::clearStandardFilters()
+{
+	EnterInitMode init;
```
same

chris-durand commented on src/modm/platform/can/sam-mcan/can_impl.hpp:
```diff
@@ -0,0 +1,619 @@
+void
+McanDriver<id, mrc>::clearExtendedFilters()
+{
+	EnterInitMode init;
```
same

chris-durand commented on src/modm/platform/can/sam-mcan/can_impl.hpp:
```diff
@@ -0,0 +1,619 @@
+void
+McanDriver<id, mrc>::configureInterrupts(uint32_t interruptPriority)
+{
+
```
```suggestion
```

chris-durand commented on src/modm/platform/can/sam-mcan/can_interrupt.cpp.in:
```diff
@@ -0,0 +1,38 @@
+#include "can_{{ id }}.hpp"
+
+
+namespace {
+using ErrorCallback = void (*)();
+}
```
same as above

chris-durand commented on src/modm/platform/can/sam-mcan/message_ram.hpp:
```diff
@@ -0,0 +1,337 @@
+
+constexpr MessageRamConfig defaultMessageRamConfig
+{
+	.filterCountStandard 	= 1,
+	.filterCountExtended	= 1,
```
Default is one filter? That is not a lot.

rleh replied:
How many filters would you suggest?
The examples in modm use at most 1 standard and one extended filter...

chris-durand replied:
My actual SAMV70 production code receives 11 different ids which can be expressed with 4 filters in that particular case. Those SAM Cortex-M7 devices all have plenty of RAM. Having 4-8 filters by default certainly wouldn't hurt much. Trying to use more filters than available silently fails at runtime. That can take a long time to debug if you haven't read the docs carefully.

rleh replied:
The `setFilter[…]()` functions return false if the filter index is out of range, but the user code needs to check for the return value...

I'll increase the number of filters to 8 each.

chris-durand commented on src/modm/platform/can/sam-mcan/message_ram.hpp:
```diff
@@ -0,0 +1,337 @@
+	.rxFifo1ElementSize		= 2*4 + modm::can::Message::capacity,
+	.rxBufferElements		= 1,
+	.rxBufferElementSize	= 2*4 + modm::can::Message::capacity,
+	.txEventFifoEntries		= 1,
```
Can this be zero? I don't think we are using the event fifo.

chris-durand reviewed (commented):
Very nice!

salkinium commented on src/modm/platform/can/sam-mcan/can.hpp:
```diff
@@ -0,0 +1,434 @@
+			// necessary workaround for devices without MCAN1 peripheral
+			return nullptr;
+			#endif
+		} else
```
```suggestion
#ifdef MCAN1
		} else if constexpr (id == 1)
 		{
 			return MCAN1;
#endif
 		} else
```

salkinium commented on src/modm/platform/can/sam-mcan/can.hpp:
```diff
@@ -0,0 +1,434 @@
+			int deadlockPreventer = 10'000; // max ~10ms
+			while (((Regs()->MCAN_CCCR & MCAN_CCCR_INIT) == 0) and (deadlockPreventer-- > 0)) {
+				using namespace std::literals;
+				modm::delay(1us);
```
```suggestion
				modm::delay_us(1);
```

salkinium commented on src/modm/platform/can/sam-mcan/can_impl.hpp:
```diff
@@ -0,0 +1,619 @@
+void
+McanDriver<id, mrc>::readMsg(modm::can::Message& message, uint8_t fifoIndex, uint8_t* filter_id, uint16_t *timestamp)
+{
+	using namespace modm::platform;
```
```suggestion
```

salkinium commented on src/modm/platform/can/sam-mcan/can_impl.hpp:
```diff
@@ -0,0 +1,619 @@
+bool
+McanDriver<id, mrc>::sendMsg(const modm::can::Message& message)
+{
+	using namespace modm::platform;
```
```suggestion
```

salkinium commented on src/modm/platform/can/sam-mcan/can_impl.hpp:
```diff
@@ -0,0 +1,619 @@
+
+	// Make sure data is written to RAM before transfer is started.
+	// Otherwise messages are corrupted.
+	__DMB();
```
No, you… you're __DMB!1!!

salkinium commented on src/modm/platform/can/sam-mcan/can_impl.hpp:
```diff
@@ -0,0 +1,619 @@
+		#ifdef MCAN1
+		ClockGen::enable<ClockPeripheral::Can1>();
+		#endif
+	}
```
```suggestion
#ifdef MCAN1
	else if constexpr (id == 1) {

 		ClockGen::enable<ClockPeripheral::Can1>();
 	}
#endif
```

salkinium replied:
there are a few more of these in lines below. It's not super important.

salkinium commented on src/modm/platform/can/sam-mcan/can_impl.hpp:
```diff
@@ -0,0 +1,619 @@
+		Regs()->MCAN_RXESC |= MCAN_RXESC_F0DS_64_BYTE;
+	}
+	else {
+		static_assert(always_false_v<>, "Invalid MessageRam::RxFifo0ElementSize value");
```
ohhhh, that's what the `always_false_v` is for. I think [this was finally fixed in C++23](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2593r0.html) 🙄

rleh replied:
> I think [this was finally fixed in C++23](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2593r0.html) roll_eyes

Nice!

But apparently it does not work yet with GCC (12.2) 😞

salkinium commented on src/modm/platform/can/sam-mcan/can_impl.hpp:
```diff
@@ -0,0 +1,619 @@
+McanDriver<id, mrc>::setMode(Mode mode)
+{
+	EnterInitMode init;
+	Regs()->MCAN_CCCR |= MCAN_CCCR_CCE;
```
isn't this already called in the [EnterInitMode constructor](https://github.com/modm-io/modm/pull/955/files#diff-2d4bc0a91ed99d7523ef1569051213915baea34e02c5f8488220c85fbf8dca23R393)?

salkinium commented on src/modm/platform/can/sam-mcan/can_impl.hpp:
```diff
@@ -0,0 +1,619 @@
+{
+	if (retransmission) {
+		// Enable retransmission
+		Regs()->MCAN_CCCR = (Regs()->MCAN_CCCR & ~MCAN_CCCR_DAR);
```
```suggestion
		Regs()->MCAN_CCCR &= ~MCAN_CCCR_DAR;
```

salkinium commented on src/modm/platform/can/sam-mcan/can_impl.hpp:
```diff
@@ -0,0 +1,619 @@
+		Regs()->MCAN_CCCR = (Regs()->MCAN_CCCR & ~MCAN_CCCR_DAR);
+	} else {
+		// Disable retransmission
+		Regs()->MCAN_CCCR = (Regs()->MCAN_CCCR | MCAN_CCCR_DAR);
```
```suggestion
		Regs()->MCAN_CCCR |= MCAN_CCCR_DAR;
```

salkinium reviewed (commented):
Very well written, just minor nitpicks.

salkinium committed:
[scripts] Use ThreadPool to compile examples

salkinium committed:
[build] Do not use gcc-13 for hosted

rleh commented:
Merge now? 😆

chris-durand commented:
> Merge now? :laughing:

I'd like more CAN filters :sweat_smile:, but apart from that, feel free to merge.

salkinium commented:
(just cherry-picked some commits to fix the macOS CI)

chris-durand reviewed (approved):
Thanks!
